sdk/python/tests/integration/e2e/test_universal_e2e.py

```diff
@@ -3,27 +3,49 @@
 from typing import Optional
 
 import pandas as pd
+import pytest
 from pytz import utc
 
 from feast import FeatureStore, FeatureView
+from tests.conftest import DataSourceCache
+from tests.data.data_creator import create_dataset
 from tests.integration.feature_repos.test_repo_configuration import (
-    Environment,
-    parametrize_e2e_test,
+    FULL_REPO_CONFIGS,
+    construct_test_environment,
+    vary_infer_feature,
 )
 from tests.integration.feature_repos.universal.entities import driver
 from tests.integration.feature_repos.universal.feature_views import driver_feature_view
 
 
-@parametrize_e2e_test
-def test_e2e_consistency(test_environment: Environment):
-    fs, fv = (
-        test_environment.feature_store,
-        driver_feature_view(test_environment.data_source),
-    )
-    entity = driver()
-    fs.apply([fv, entity])
+@pytest.mark.integration
+@pytest.mark.parametrize(
+    "config", vary_infer_feature(FULL_REPO_CONFIGS), ids=lambda v: str(v)
+)
+def test_e2e_consistency(config, data_source_cache: DataSourceCache):
+    df = create_dataset()
+    test_suite_name = "test_e2e_consistency"
+    with construct_test_environment(test_suite_name, config) as test_environment:
+        fs = test_environment.feature_store
+        infer_features = test_environment.test_repo_config.infer_features
+        key = f"{test_suite_name}_{test_environment.test_repo_config.offline_store_creator}"
+        _, _, data_source, _ = data_source_cache.get_or_create(
```

Comment:
> Can you expand on that you think the fixture should contain? Should it just be the dataframes? Or also data sources?

It would only return environments.

It's possible to do something like

```
environment_configs = [
    SomeEnvConfigThing(provider="local"),
    SomeEnvConfigThing(provider="gcp"),
    SomeEnvConfigThing(provider="aws"),
]


@pytest.fixture(params=environment_configs, scope="module")
def environment(request):
    e = Environment(config=request.param)
    e = stage_data(e)
    e = apply(e)
    e = materialize(e)
    yield e
    e.teardown()
```

which could then be accessed like

```
def test_online_retrieval(environment):
    # do stuff
```

`test_online_retrieval` would be called once per environment. We could then also do things like

````
@pytest.mark.parametrize('some_other_flag', [True, False])
def test_online_retrieval(environment, some_other_flag):
    # do stuff
```

Which would run `test_online_retrieval` twice per env, with `some_other_flag` toggled on and off. But importantly the environment is only created once.

Another approach is to do something like

```
@pytest.fixture()
def bigquery_universal_datasource():
    from universal_repo import driver_df, driver_fv, customer_df, customer_fv

    dsc = BigQueryDataSourceCreator()
    dsc.create_data_source(driver_df, driver_fv)
    dsc.create_data_source(customer_df, customer_fv)
    yield dsc
    dsc.teardown()


@pytest.mark.parametrize(
    "provider,data_source,online_store,full_feature_names",
    [
        (
            "local",
            pytest.lazy_fixture("local_universal_datasource"),
            "sqlite",
            True
        ),
        (
            "local",
            pytest.lazy_fixture("local_universal_datasource"),
            "sqlite",
            False
        ),
        (
            "local",
            pytest.lazy_fixture("local_universal_datasource"),
            "redis",
            True
        ),
        (
            "local",
            pytest.lazy_fixture("local_universal_datasource"),
            "dymamodb",
            True
        ),
        (
            "local",
            pytest.lazy_fixture("local_universal_datasource"),
            "datastore",
            True
        ),
        (
            "gcp",
            pytest.lazy_fixture("bigquery_universal_datasource"),
            "sqlite",
            True
        ),
        (
            "aws",
            pytest.lazy_fixture("redshift_universal_datasource"),
            "sqlite",
            True
        )
    ],
)
def test_online_retrieval(provider, data_source, online_store, full_feature_names):
    with Environment(config(provider, data_source, online_store, full_feature_names) as e:
        # do stuff
```
The above API isn't functional, but the idea is that we'd use `lazy_fixtures` and scope the fixtures to session/module/function depending on if we want isolation or reuse across all tests. So we're essentially leveraging the caching that is built into Pytest since it won't redo the staging of data, and it's 100% clear what the test cases for a specific test are.

And if you as a test creator need to do anything off the beaten path then you can set up your own DataSourceCreator and feed it whatever datasets you want (and use a different repo than universal repo)

Reply:
> I think with the universal tests, there's many different tests that operate on what is basically the same data. So even if the test functions are different, there's advantages in sharing the data sources - for me it's equivalent to having the test data uploaded to redshift/bigquery outside of the tests and used directly with the assumptions that they exist. We would be sharing the same data, without needing different data sets for different tests.

100% agree with solving this problem, just want to keep things as simple as possible.

> There is an issue where both universal and non-universal sources are cached in the same container - which I can see value in keeping separate. Maybe I can create a different cache that's session scoped for universal tests, and function scoped for non-universal tests?

Well if we go with the cache approach, isn't it better to give control to the test writer? They can use keys to purposefully select or miss existing cache items, as long as the cache keys are named according to some standard.